### DISCUSSION TITLE
DllNotFoundException on .net 6 with python 3.12

### DISCUSSION THREAD

**Discussion by software6ptk** [pythonnet/pythonnet#2340] [Q&A]
### Environment

-   Pythonnet version: 3.0.3
-   Python version: 3.12
-   Operating System: Windows 11 pro
-   .NET Runtime: 6

### Details

I am trying to setup pythonnet to run some python scripts in my c# app, but i'm getting the usual DllNotFoundException when setting the Runtime.PythonDLL variable

If i'm not wrong, both the python version and the .net runtime are supported so that shouldn't be the issue
![image](https://github.com/pythonnet/pythonnet/assets/141343509/4a0e4e24-b161-4bac-aaec-5e40d5b9f010)
![image](https://github.com/pythonnet/pythonnet/assets/141343509/9904b2da-6dc5-4045-b411-4ae189e7f948)

this is what i do
```c#
    
            string dllpth = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
            dllpth += "\\Local\\Programs\\Python\\Python312\\python312.dll";

            Runtime.PythonDLL = dllpth;
            PythonEngine.Initialize();

```

i also tried

```c#
    
            string dllpth = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
            dllpth += "\\Local\\Programs\\Python\\Python312\\python312.dll";

            Environment.SetEnvironmentVariable("PYTHONNET_PYDLL", dllpth);
            PythonEngine.Initialize();

```

Errors are:

```
System.TypeInitializationException: 'The type initializer for 'Delegates' threw an exception.'
Win32Exception: Impossibile trovare il modulo specificato.
```

and

```
System.TypeInitializationException: 'The type initializer for 'Delegates' threw an exception.'
DllNotFoundException: Could not load C:\Users\***\AppData\Roaming\Local\Programs\Python\Python312\python312.dll.
```

with details:

```
System.TypeInitializationException
  HResult=0x80131534
  Messaggio=The type initializer for 'Delegates' threw an exception.
  Origine=Python.Runtime
  Analisi dello stack:
   in Python.Runtime.Runtime.<>c.<Initialize>b__32_0()
   in Python.Runtime.Runtime.TryUsingDll[T](Func`1 op)
   in Python.Runtime.Runtime.Initialize(Boolean initSigs)
   in Python.Runtime.PythonEngine.Initialize(IEnumerable`1 args, Boolean setSysArgv, Boolean initSigs)
   in GeometriesVisualizer_EYESHOT.ScriptManager.RunLoaded(Object parameter) in C:\Users\***\ScriptManager.cs: riga 146
   in GeometriesVisualizer_EYESHOT.MainWindow..ctor() in C:\Users\***\MainWindow.xaml.cs: riga 124
   in System.RuntimeType.CreateInstanceDefaultCtor(Boolean publicOnly, Boolean wrapExceptions)

Eccezione interna 1:
DllNotFoundException: Could not load C:\Users\***\AppData\Roaming\Local\Programs\Python\Python312\python312.dll.

Eccezione interna 2:
Win32Exception: Impossibile trovare il modulo specificato.
```

I already tried:
- reinstalling python
- changing build target to x64

I also searched online for a solution, but can't find anything besides changing the build target

Can you please help me?

**Comment by izhgtumekhanik** [comment]
For Python x64 use x64 profile app (Any CPU - not working)

**Comment by DaveSkender** [comment]
Is there a chance you're just not installing the package correctly?

```bash
# from your .csproj directory
dotnet add package pythonnet
```

Or possibly Python itself?  Sounds like Python wasn't added to your environment PATH variables.  Our [Stock Indicators for Python QuickStart](https://github.com/facioquo/stock-indicators-python-quickstart) may help with more details on how to setup your environment, though our use case is a bit different.

**Comment by Gianmaria92** [comment]
Hey, @software6ptk , did you solve the problem?

**Comment by mpulukkinen** [comment]
I'm having ecapt same issue. Weird thing is that on my dev computer, I got it working. Then I updated my app's Flight in MS store and tested with my laptop: getting the dllNotFound whatever I do. Tried several different versions and also checked that maybe the python38.dll (or 310.dll) has some dependency missing, but no...

**Reply by mpulukkinen** [reply]
Actually, my issue was that one of the dependencies, msvc was not installed in target machine. I discovered when I ditched the pythonnet and starter running python just with process.start. So, it was actually one of the python scripts that failed to load the dll, but pythonnet completely "ate" the original error and made it look like the pythonnet.dll had some missing dependency. Pretty common case with c# and dependency dll load issues: the actual root cause gets lost in translation 😆

**Comment by sharafabacery** [comment]
```
static void Main(string[] args)
{

    Runtime.PythonDLL = @"python313.dll";
    PythonEngine.Initialize();

    using (Py.GIL())
    {
        // Import TensorFlow preprocessing script
        string modulePath = @"D:\ocr_v1\ConsoleApp1\ConsoleApp1\AiNet";
        dynamic sys = Py.Import("sys");
        sys.path.append(modulePath);

        // Import your Python module and use it
        dynamic utils = Py.Import("utils");
        // Process image and run model prediction
        string imagePath = "D:\\ocr_v1\\ConsoleApp1\\ConsoleApp1\\AiNet\\20160805-130.webp";
        dynamic result = utils.detect_and_process_id_card(imagePath);

        Console.WriteLine($"Predicted Class: {result}");
    }

    PythonEngine.Shutdown();
}
```
I used something like that and worked to me and I create configuration manger of X64 to be combatable with with my vs and here where python live `C:\Users\User\AppData\Local\Programs\Python\Python313`

**Reply by sharafabacery** [reply]
![Capture15](https://github.com/user-attachments/assets/409c682a-21f9-4f25-ad58-5b5da7691147)
